### PR TITLE
SSCS-10035 correction refused

### PR DESCRIPTION
**Pull request by nickhill111**
https://tools.hmcts.net/jira/browse/SSCS-10035

Contains all of SSCS-10523

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/IssueDocumentHandler.java [hunk 1]
```diff
@@ -210,4 +209,24 @@ protected Optional<String> buildAppointeeName(SscsCaseData caseData) {
...
+        if (nonNull(postHearingReviewType)) {
+            switch (postHearingReviewType) {
+                case SET_ASIDE:
+                    return caseData.getDocumentGeneration().getBodyContent();
+                case CORRECTION:
+                    return caseData.getDocumentGeneration().getCorrectionBodyContent();
+                case STATEMENT_OF_REASONS:
+                case PERMISSION_TO_APPEAL:
+                case LIBERTY_TO_APPLY:
+                default:
```
Comment (amoslv1) [inline src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/IssueDocumentHandler.java:225]: I'd like to see an exception thrown here. Guards against accidental omission when the remaining post hearing types are implemented. See some example methods in PdfRequestUtil.

File: src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/posthearingreview/PostHearingReviewSubmittedHandler.java [hunk 2]
```diff
@@ -83,7 +84,7 @@ private void handlePostHearingRefused(SscsCaseData caseData) {
         Long caseId = Long.valueOf(caseData.getCcdCaseId());
         PostHearing postHearing = caseData.getPostHearing();
         if (postHearing.isRefused()) {
-            if (isYes(postHearing.getSetAside().getRequestStatementOfReasons())) {
+            if (SET_ASIDE.equals(postHearing.getReviewType()) && isYes(postHearing.getSetAside().getRequestStatementOfReasons())) {
```
Comment (amoslv1) [inline src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/posthearingreview/PostHearingReviewSubmittedHandler.java:87]: Is this additional check strictly necessary? If the type is not set aside, then getSetAside() will generate a new blank SetAside, in which getRequestStatementOfReasons will be null.

File: src/test/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/posthearingreview/PostHearingReviewAboutToSubmitHandlerTest.java [hunk 3]
```diff
@@ -94,6 +82,24 @@ void shouldReturnWithoutError() {
...
+        caseData.getPostHearing().getCorrection().setAction(CorrectionActions.REFUSE);
+
+        PreSubmitCallbackResponse<SscsCaseData> response =
+            handler.handle(ABOUT_TO_SUBMIT, callback, USER_AUTHORISATION);
+
+        assertThat(response.getErrors()).isEmpty();
+
+        List<SscsDocument> documents = response.getData().getSscsDocument();
+        assertThat(documents).hasSize(1);
+        assertThat(documents.get(0).getValue().getDocumentType()).isEqualTo(DocumentType.CORRECTION_REFUSED.getValue());
```
Comment (amoslv1) [inline src/test/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/posthearingreview/PostHearingReviewAboutToSubmitHandlerTest.java:100]: extracting can be used to chain assertions here
```suggestion
assertThat(documents)
    .hasSize(1)
    .element(0)
    .extracting(doc -> doc.getValue().getDocumentType())
    .isEqualTo(DocumentType.CORRECTION_REFUSED.getValue());
```

**Review by amoslv1 — CHANGES_REQUESTED** [review]
Some code optimisations

File: src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/posthearingreview/PostHearingReviewSubmittedHandler.java (hunk 2 above)
Comment (nickhill111) [reply src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/posthearingreview/PostHearingReviewSubmittedHandler.java:87]: Adding this check for clarity and so i can push this ticket forward whilst i investigate the bug we had a call about

File: src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/IssueDocumentHandler.java (hunk 1 above)
Comment (nickhill111) [reply src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/IssueDocumentHandler.java:225]: done

File: src/test/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/IssueDocumentHandlerTest.java [hunk 4]
```diff
@@ -195,4 +199,24 @@ public void givenHearingIsNull_thenReturnDraftDecisionNotice() {
...
+        String bodyContent = "correction body content";
+        sscsCaseData.getDocumentGeneration().setCorrectionBodyContent(bodyContent);
+
+        NoticeIssuedTemplateBody payload = handler.createPayload(null, sscsCaseData, "doctype", LocalDate.now(), LocalDate.now(), false, USER_AUTHORISATION);
+
+        assertEquals(payload.getNoticeBody(), bodyContent);
+    }
+
+    @Test
+    public void givenPostHearingReviewIsLibertyToApply_thenThrowException() {
```
Comment (amoslv1) [inline src/test/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/IssueDocumentHandlerTest.java:215]: I would change this to ParameterizedTest and call it givenPostHearingReviewTypeIsNotImplemented_thenThrowException. Make the input PostHearingReviewType.class with the names of the three yet to be implemented.

File: src/test/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/IssueDocumentHandlerTest.java [hunk 5]
```diff
@@ -195,4 +199,24 @@ public void givenHearingIsNull_thenReturnDraftDecisionNotice() {
...
+        SscsCaseData sscsCaseData = buildCaseData();
+        sscsCaseData.getPostHearing().setReviewType(PostHearingReviewType.LIBERTY_TO_APPLY);
+
+        assertThrows(IllegalArgumentException.class, () ->
+            handler.createPayload(null, sscsCaseData, "doctype", LocalDate.now(), LocalDate.now(), false, USER_AUTHORISATION));
```
Comment (amoslv1) [inline src/test/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/IssueDocumentHandlerTest.java:220]: You can chain an expected message on the end as well, .hasMessage or .hasMessageStartingWith
May need to use a different Assertion method like assertThrownBy.
```suggestion
        assertThrownBy(() ->
            handler.createPayload(null, sscsCaseData, "doctype", LocalDate.now(), LocalDate.now(), false, USER_AUTHORISATION))
            .hasType(IllegalArgumentException.class)
            .hasMessage("...");
```

**Review by amoslv1 — CHANGES_REQUESTED** [review]
couple of final tweaks

File: src/test/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/IssueDocumentHandlerTest.java (hunk 5 above)
Comment (nickhill111) [reply src/test/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/IssueDocumentHandlerTest.java]: done

**Review by amoslv1 — DISMISSED** [review]

**Pull request closed without merging**